### PR TITLE
Sparse matrix addition (related to Trilinos pull request #1957)

### PR DESCRIPTION
**Pull request by brian-kelley**
Same interface as spgemm (separate symbolic/numeric)
Implemented in sparse/KokkosSparse_spadd.hpp
Has associated handle type (like spgemm) and unit tests

### CODE REVIEW

File: src/sparse/KokkosSparse_spadd.hpp [hunk 1]
```diff
@@ -0,0 +1,733 @@
...
+        execution_space::fence();
+        //compute actual c_rowmap
+        parallel_prefix_sum<size_type, clno_row_view_t_> prefix(c_rowcounts, c_rowmap);
+        Kokkos::parallel_scan(range_type(0, nrows + 1), prefix);
+        execution_space::fence();
+      }
+      addHandle->set_a_b_pos(a_pos, b_pos);
+    }
+    //provide the number of NNZ in C to user through handle
+    addHandle->set_max_result_nnz(c_rowmap(nrows));
```
Comment (mndevec) [inline src/sparse/KokkosSparse_spadd.hpp]: This cause access issues if UVM is not enabled. Something like below is needed for this.
    auto d_c_nnz_size = Kokkos::subview(c_rowmap, nrows);
    auto h_c_nnz_size = Kokkos::create_mirror_view (d_c_nnz_size);
    Kokkos::deep_copy (h_c_nnz_size, d_c_nnz_size);
    execution_space::fence();
    addHandle->set_max_result_nnz(h_c_nnz_size());

File: src/sparse/KokkosSparse_spadd.hpp [hunk 2]
```diff
@@ -0,0 +1,733 @@
...
+        UnsortedEntriesUpperBound<size_type, alno_row_view_t_, blno_row_view_t_, clno_row_view_t_>
+          countEntries(a_rowmap, b_rowmap, c_rowcounts_upperbound);
+        Kokkos::parallel_for(range_type(0, nrows), countEntries);
+        execution_space::fence();
+        //get (temporary) c_rowmap as cumulative sum
+        parallel_prefix_sum<size_type, clno_row_view_t_> prefix(c_rowcounts_upperbound, c_rowmap_upperbound);
+        Kokkos::parallel_scan(range_type(0, nrows + 1), prefix);
+        //compute uncompressed entries of C (just indices, no scalars)
+        execution_space::fence();
+        c_nnz_upperbound = c_rowmap_upperbound(nrows);
```
Comment (mndevec) [inline src/sparse/KokkosSparse_spadd.hpp]: auto d_c_nnz_size = Kokkos::subview(c_rowmap_upperbound, nrows);
        auto h_c_nnz_size = Kokkos::create_mirror_view (d_c_nnz_size);
        Kokkos::deep_copy (h_c_nnz_size, d_c_nnz_size);
        execution_space::fence();
        c_nnz_upperbound = h_c_nnz_size();

        auto d_c_nnz_size = Kokkos::subview(c_rowmap_upperbound, nrows);
        auto h_c_nnz_size = Kokkos::create_mirror_view (d_c_nnz_size);
        Kokkos::deep_copy (h_c_nnz_size, d_c_nnz_size);
        execution_space::fence();
        c_nnz_upperbound = h_c_nnz_size();

File: src/sparse/KokkosSparse_spadd.hpp [hunk 3]
```diff
@@ -0,0 +1,733 @@
...
+      Kokkos::parallel_for(range_type(0, nrows), unmergedSum);
+      execution_space::fence();
+      //sort the unmerged sum
+      SortEntriesFunctor<size_type, clno_row_view_t_, clno_nnz_view_t_>
+        sortEntries(c_rowmap_upperbound, c_entries_uncompressed, ab_perm);
+      Kokkos::parallel_for(range_type(0, nrows), sortEntries);
+      execution_space::fence();
+      clno_nnz_view_t_ a_pos("A entry positions", a_rowmap(nrows));
+      clno_nnz_view_t_ b_pos("B entry positions", b_rowmap(nrows));
+      //merge the entries and compute Apos/Bpos, as well as Crowcounts
```
Comment (mndevec) [inline src/sparse/KokkosSparse_spadd.hpp]: same this for lines 477 and 478. The array cannot be accessed by the host.
Simple we can use  a_entries.dimension_0() here, right?

**Comment by crtrott** [comment]
Somebody needs to rework this to follow our coding guidelines as well as adding ETI and the TPL hooks etc.

**Comment by mndevec** [comment]
@crtrott 
I have the issue listed here: 
https://github.com/kokkos/kokkos-kernels/issues/151

**Pull request merged**